### PR TITLE
Robustify OPAL_WHICH

### PR DESCRIPTION
**Pull request by jsquyres**
Make OPAL_WHICH properly respond to absolute pathnames, and relative
pathnames that contain one or more path separators.  See comment in
the code explaining the three cases.

Prior to this commit, passing /path/to/some/valid/executable into
OPAL_WHICH would result in $2 not being set.

Signed-off-by: Jeff Squyres <jsquyres@cisco.com>

Refs #9388.  This fix will also need to be cherry-picked to the v5.0.x branch.  The same OPAL_WHICH problem exists in v4.0.x and v4.1.x, but may not be an issue there, because master/v5.0.x uses the output of OPAL_WHICH for the C++ compiler differently than the v4.0.x and v4.1.x branches.

### CODE REVIEW

File: config/opal_functions.m4 [hunk 1]
```diff
@@ -480,24 +480,56 @@ dnl #######################################################################
...
+    # Otherwise, $2 is unchanged.
+
+    # 3. opal_prog contains no /.  Search the PATH for an excutable
+    # with the appropriate name.  If found, return the absolute path
+    # in $2.  Otherwise, $2 is unchanged.
+
+    # Note that these three cases are exactly what which(1) does.
+
+    # Case #3: opal_prog contains no /
+    echo $opal_prog | grep -q /
```
Comment (bosilca) [inline config/opal_functions.m4:509]: Maybe should use DIRECTORY_SEPARATOR instead of the hardcoded /

File: config/opal_functions.m4 [hunk 2]
```diff
@@ -480,24 +480,56 @@ dnl #######################################################################
...
+           IFS="$PATH_SEPARATOR"
+           for opal_dir in $PATH; do
+               if test -x "$opal_dir/$opal_prog"; then
+                   $2="$opal_dir/$opal_prog"
+                   break
+               fi
+           done
+           IFS=$IFS_SAVE],
+          [ # Case #1: opal_prog is an absolute filename
+           echo $opal_prog | cut -c1 | grep -q /
```
Comment (bosilca) [inline config/opal_functions.m4:521]: same comment as above.

File: config/opal_functions.m4 (hunk 1 above)
Comment (jsquyres) [reply config/opal_functions.m4:509]: Good point.  Will fix both places.

**Review by bosilca — APPROVED** [review]

File: config/opal_functions.m4 [hunk 3]
```diff
@@ -480,24 +480,56 @@ dnl #######################################################################
...
+    # Otherwise, $2 is unchanged.
+
+    # 3. opal_prog contains no /.  Search the PATH for an excutable
+    # with the appropriate name.  If found, return the absolute path
+    # in $2.  Otherwise, $2 is unchanged.
+
+    # Note that these three cases are exactly what which(1) does.
+
+    # Case #3: opal_prog contains no /
+    echo $opal_prog | grep -q $PATH_SEPARATOR
```
Comment (bwbarrett) [inline config/opal_functions.m4:509]: I do not think this is right.  `$PATH_SEPARATOR` will be `:`, not `/`.

Comment (bosilca) [reply config/opal_functions.m4:509]: Indeed, DIRECTORY_SEPARATOR should be used here and in the Case 1 below.

**Review by bwbarrett — APPROVED** [review]

**Pull request merged**